packages/react/src/internal/usePreviousValue.ts

```diff
@@ -0,0 +1,21 @@
+/**
+ * Copyright IBM Corp. 2021, 2021
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { useEffect, useRef } from 'react';
+
+/**
+ * Returns the previous state value included in the param
+ * @param {object} value - The current value of any type.
+ * @returns {T | undefined} - The previous value of the same type, or undefined if there is none
+ */
+export const usePreviousValue = (value) => {
+  const ref = useRef(undefined);
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+};
```

Comment:
The JSDoc comment doesn't have the correct types (the comment doesn't even need the types as long as the function includes them) and the function is missing generics.

```suggestion
/**
 * Stores the previous value of a given input.
 *
 * @param value - The current value.
 * @returns The value before the current render.
 *
 * @example
 * const [count, setCount] = useState(0);
 * const prevCount = usePreviousValue(count);
 */
export const usePreviousValue = <T,>(value: T): T | undefined => {
  const ref = useRef<T | undefined>(undefined);

  useEffect(() => {
    ref.current = value;
  }, [value]);

  return ref.current;
};
```

Reply:
Make sense.  (I just copied the code from @carbon/ibm-products, but this is an improvement.)

Reply:
Maybe their code needs to be updated too.